### PR TITLE
Whisper output normalization

### PR DESCRIPTION
**Pull request by Leili**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Adding whisper output normalization for GRPO 
Normalization currently covers following languages: english, spanish, french, german, mandarin  
If no language is specified, it uses english by default 

**Collection**: [Note which collection this PR will affect]

TTS

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/tts/models/magpietts_preference_optimization.py [hunk 1]
```diff
@@ -60,6 +72,8 @@ def __init__(self, cfg: DictConfig, trainer: 'Trainer' = None):
             self.whisper_model = WhisperForConditionalGeneration.from_pretrained("openai/whisper-large-v3")
             self.whisper_model.eval()
 
+        self._normalize_whisper_transcript = cfg.get('normalize_whisper_transcript', True)
```
Comment (blisc) [inline nemo/collections/tts/models/magpietts_preference_optimization.py]: You define _normalize_whisper_transcript in the MagpieTTSModelOfflinePODataGen class but you don't initialize them in this class. Can you add an initialization to this class?

File: nemo/collections/tts/models/magpietts_preference_optimization.py [hunk 2]
```diff
@@ -37,6 +37,18 @@
     HAVE_TORCHAUDIO = False
 
 from nemo.collections.tts.models import MagpieTTSModel
+from nemo_text_processing.text_normalization.normalize import Normalizer
+
+# Global cache for normalizers - to avoid latency from repeated creation
+_NORMALIZER_CACHE = {}
```
Comment (blisc) [inline nemo/collections/tts/models/magpietts_preference_optimization.py]: Instead of creating a global variable, can you create it as part of the classes? And then pass the normalizer object to the transcribe_with_whisper function instead of a boolean? Something like
```
def transcribe_with_whisper(audio_filepath, language, whisper_processor, whisper_model, device, normalizer: Optional[Normalizer] = None):
   ...
   if normalizer is not None:
      ...
```

File: nemo/collections/tts/models/magpietts_preference_optimization.py (hunk 1 above)
Comment (Leili) [reply nemo/collections/tts/models/magpietts_preference_optimization.py]: Hi @blisc 
Can you help me understand this comment? I'm initializing the variable to True, if it's not found in the config

File: nemo/collections/tts/models/magpietts_preference_optimization.py (hunk 2 above)
Comment (shehzeen) [reply nemo/collections/tts/models/magpietts_preference_optimization.py]: looks like this has been addressed, and the normalizer cache is now initialized within the class. Marking this as resolved.

File: nemo/collections/tts/models/magpietts_preference_optimization.py (hunk 1 above)
Comment (shehzeen) [reply nemo/collections/tts/models/magpietts_preference_optimization.py]: _normalize_whisper_transcript is being initialized inside the `if cfg.get('load_whisper_model', False):` and is also being used in a similar if statement in the test_step. It should be fine. @blisc  is your concern that we might end up using it elsewhere in the code and we should avoid that?

**Comment by blisc** [comment]
Will merge this, but will use #14600 instead as we are updating the dev branches
